[SRVKS-187] Add telemetry support

skonto opened the pull request:
- Monitoring is failing on master. Fixes the following error:
```
{"level":"error","ts":"2020-11-02T10:27:53.126Z","logger":"cmd","msg":"Failed to start monitoring","error":"the cache is not started, can not read objects","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/github.com/go-logr/zapr/zapr.go:132\nmain.main\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/cmd/manager/main.go:110\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203"}
```
The recommendation at this early stage is to use our own [client](https://github.com/jaegertracing/jaeger-operator/issues/809#issue-532751170).

- Adds telemetry support according to  SRVKS-187. Metrics exposed are shown next:
![image](https://user-images.githubusercontent.com/7945591/97861708-da0c8a00-1d0c-11eb-84a7-dc79d34525d6.png)
For the above example here are the actual resources:
```
$ oc get revisions -n knative-serving
NAME           CONFIG NAME   K8S SERVICE NAME   GENERATION   READY   REASON
kn-cli-r5vr2   kn-cli        kn-cli-r5vr2       1            True    
$ oc get revisions
NAME                  CONFIG NAME     K8S SERVICE NAME      GENERATION   READY   REASON
event-display-8tdww   event-display   event-display-8tdww   1            True    
 $ oc get revisions -n apisource
NAME                  CONFIG NAME     K8S SERVICE NAME      GENERATION   READY   REASON
event-display-nphxs   event-display   event-display-nphxs   1            True    

```

Note: PR i splitted in different commits for easier review. You can skip vendor.

/cc @markusthoemmes @matzew @s-urbaniak @openshift-team-monitoring

markusthoemmes commented on knative-operator/hack/update-deps.sh:
```diff
@@ -20,6 +20,7 @@ OCP_VERSION="release-4.6"
 FLOATING_DEPS=(
   "github.com/openshift/api@${OCP_VERSION}"
 
+  "knative.dev/eventing@${KN_VERSION}"
```
please add eventing-contrib too as it's now part of go.mod as well.

skonto replied:
yeah sure.

markusthoemmes commented:
> Monitoring is failing on master. Fixes the following error:

Please send a separate PR just for this fix. Can we please add E2E tests of some sorts that would catch something like this there too?

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,122 @@
+		return err
+	}
+	for _, tp := range typesAndMetrics {
+		if err := c.Watch(&source.Kind{Type: tp}, &handler.EnqueueRequestForObject{}, skipUpdatePredicate{}, updateMetricsDeletePredicate{}, updateMetricsCreatePredicate{}); err != nil {
```
Seems like we only need the watch functionality? Can we add that in a different file that doesn't look like a controller to make that a bit more obvious?

skonto replied:
You mean in an existing file?

markusthoemmes replied:
No necessarily, but if we are not going to make this into a controller, let's not make it look like one.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (updateMetricsCreatePredicate) Create(e event.CreateEvent) bool {
+	matchAndUpdateMetric(e.Object, inc)
```
I'd change this a bit to be more pluggable, like:

```
metric := getMetricFor(e.Object)
metric.Inc()
```

This makes the `op` type unnecessary and composes better imo.

skonto commented:
@markusthoemmes yes that is my plan to add integration tests. We had this but due to lack of readiness probe they were not reliable. I will need to add the probe thingy first.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_metrics.go:
```diff
@@ -0,0 +1,25 @@
+		},
+		[]string{"type"},
+	)
+	ServicesG  = serverlessTelemetryG.WithLabelValues("service")
+	RoutesG    = serverlessTelemetryG.WithLabelValues("route")
+	RevisionsG = serverlessTelemetryG.WithLabelValues("revision")
+	SourcesG   = serverlessTelemetryG.WithLabelValues("source")
```
How do these differ from the metrics that we surface via `GenerateAndServeCRMetrics`?

skonto replied:
I tried that framework before coming up with this simplified solution. Here are the downsides:
- It requires a cfg object passed by a controller which makes it hard to test.
- it adds a redundant label for the CR type that distinguishes the type of metric [here](https://github.com/operator-framework/operator-sdk/blob/125d0dfcc71fef4f9d7e2a42b1354cb79ffdee03/pkg/kube-metrics/metrics.go#L100), increasing the carnality (max is 10 for telemetry).

markusthoemmes commented:
@skonto you can investigate the new c-r implementations regarding this. I believe they have "native" readiness endpoint support, but I haven't yet looked into that.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,122 @@
+		return err
+	}
+	for _, tp := range typesAndMetrics {
+		if err := c.Watch(&source.Kind{Type: tp}, &handler.EnqueueRequestForObject{}, skipUpdatePredicate{}, updateMetricsDeletePredicate{}, updateMetricsCreatePredicate{}); err != nil {
```
Can we fold all the predicates into just one handler?

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,122 @@
+
+// ReconcileResourcesForTelemetry reconciles a source deployment object
+type ReconcileResourcesForTelemetry struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
```
Are these used?

markusthoemmes replied:
Most likely you won't need this one.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,122 @@
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("telemetry-resources-discovery-controller", mgr, controller.Options{Reconciler: r})
```
What happens here if we don't provide a Reconciler in the options?

markusthoemmes replied:
nvm, throws an error.

skonto replied:
Let me check to see how it will behave.

skonto replied:
@markusthoemmes the most limited example I see is [this](https://github.com/kubernetes-sigs/controller-runtime/blob/af24f3b6f46793c3210224bdd00c2593970942f1/pkg/controller/example_test.go#L44-L48) one.

markusthoemmes replied:
Nice, yeah let's make use of that!

skonto replied:
Reconciler [needs](https://github.com/kubernetes-sigs/controller-runtime/blob/af24f3b6f46793c3210224bdd00c2593970942f1/pkg/controller/controller.go#L92) to be specified.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,85 @@
+	return true
+}
+
+func getMetricFor(obj runtime.Object) *prometheus.Gauge {
```
Why return a pointer if we dereference it above anyway? :thinking:

skonto replied:
It covers the case where the objectype is not the one we want.

skonto replied:
Can be simplified though its a struct anyway, fixed it.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller.go:
```diff
@@ -0,0 +1,85 @@
+	if metric := getMetricFor(e.Object); metric != nil {
+		(*metric).Dec()
+	}
+	return true
```
I think we can just return false everywhere, so the reconciler is just never called.

skonto replied:
Ok

skonto commented:
I created 1000 sources at once in order to see the behavior of the telemetry watcher:
![image](https://user-images.githubusercontent.com/7945591/98136418-9448f080-1ec9-11eb-99e3-fb0a16783715.png)
I didnt notice any issues. Finally the metric was updated to the right value.

skonto commented:
@markusthoemmes I updated the PR but see errors [here](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-knative_serverless-operator/657/pull-ci-openshift-knative-serverless-operator-master-4.6-upstream-e2e-aws-ocp-46/1324020938789883904/artifacts/upstream-e2e-aws-ocp-46/gather-extra/pods/openshift-serverless_knative-openshift-ffb8c95f-6lp87_knative-openshift.log).
```
controller","error":"failed to wait for health-controller caches to sync: cache did not sync","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/github.com/go-logr/zapr/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/sigs.k8s.io/controller-
```
The resources are added to the scheme. This is something I didnt see locally so probably c-r is not capable to handle CRs not yet installed.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+	}}
+	ret = []metricCase{}
+	for _, v := range objects {
+		ret = append(ret, metricCase{
```
nit: You can make this all one `append` call, like

```golang
ret = append(ret, metricCase{
  name: fmt.Sprintf("create a %s", v.name),
  event: event.CreateEvent{Object: v.obj},
  expectedMetricValue: 1,
}, metricCase{
  name: fmt.Sprintf("create a %s", v.name),
  event: event.CreateEvent{Object: v.obj},
  expectedMetricValue: 1,
})

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+var (
+	service           = servingv1.Service{}
+	route             = servingv1.Route{}
+	revision          = servingv1.Revision{}
+	pingSource        = eventingsourcesv1beta1.PingSource{}
+	apiServerSource   = eventingsourcesv1beta1.ApiServerSource{}
+	SinkBindingSource = eventingsourcesv1beta1.SinkBinding{}
+	kafkaSource       = kafkasourcev1beta1.KafkaSource{}
```
If you make all of those pointers, you won't need the `DeepCopyObject` calls below, as they'll then fit into the `runtime.Object` interface themselves.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+func generateMetricUpdateSteps() (ret []metricCase) {
+	objects := []struct {
+		name string
+		obj  runtime.Object
+	}{{
+		name: "service",
+		obj:  service.DeepCopyObject(),
+	}, {
+		name: "route",
+		obj:  route.DeepCopyObject(),
+	}, {
+		name: "revision",
+		obj:  revision.DeepCopyObject(),
+	}, {
+		name: "pingsource",
+		obj:  pingSource.DeepCopyObject(),
+	}, {
+		name: "apiserversource",
+		obj:  apiServerSource.DeepCopyObject(),
+	}, {
+		name: "sinkbindingsource",
+		obj:  SinkBindingSource.DeepCopyObject(),
+	}, {
+		name: "kafkasource",
+		obj:  kafkaSource.DeepCopyObject(),
+	}}
```
nit: You could replace this with globally using a map, i.e:

```golang
objects := map[string]runtime.Object{
  "service": &servingv1.Service{}
  ...
}
```

I'd even move the entire initializiation here if you don't need the objects globally.

skonto replied:
Initiall I had map but iterating on it didnt preserve the order.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+	revision          = servingv1.Revision{}
+	pingSource        = eventingsourcesv1beta1.PingSource{}
+	apiServerSource   = eventingsourcesv1beta1.ApiServerSource{}
+	SinkBindingSource = eventingsourcesv1beta1.SinkBinding{}
```
Why export?

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+		mp := metricsPredicate{}
+		dto := ioprometheusclient.Metric{}
+		var metric prometheus.Gauge
+		switch tc.event.(type) {
+		case event.CreateEvent:
+			mp.Create(tc.event.(event.CreateEvent))
+			metric = getMetricFor(tc.event.(event.CreateEvent).Object)
+		case event.DeleteEvent:
+			mp.Delete(tc.event.(event.DeleteEvent))
+			metric = getMetricFor(tc.event.(event.DeleteEvent).Object)
+		case event.UpdateEvent:
+			mp.Update(tc.event.(event.UpdateEvent))
+			metric = getMetricFor(tc.event.(event.UpdateEvent).ObjectOld)
+		}
```
nit: Consider rewriting this with type assertions maybe to avoid the "double casting", like:

```golang
if create, ok := tc.event.(event.CreateEvent); ok {
  mp.Create(create)
  metric = getMetricFor(create.Object)
} else if delete, ok := tc.event.(event.DeleteEvent); ok {
 ...
}
```

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_controller_test.go:
```diff
@@ -0,0 +1,108 @@
+			t.Fatal("Cannot write metric:", err)
+		}
+		if *dto.Gauge.Value != tc.expectedMetricValue {
+			t.Errorf("Got = %v, want: %v for event: %v", *dto.Gauge.Value, tc.expectedMetricValue, n)
```
Is the `n` really helpful here? Seems like we'd actually want to print `tc` maybe?

markusthoemmes reviewed (commented):
Mostly nits to make code a bit less repetitive. Looks good otherwise!

skonto commented:
I will try to use polling instead to avoid the c-r limitations.

skonto commented:
@markusthoemmes I updated the PR addressing your comments but also covering the failing case where CRD was not found (didnt go for the polling solution because of the cpu overhead it may cause). the new design adds an unmanaged controller per component following the example [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/example_test.go).
I cover the lifecycle of each component by starting and stopping the corresponding controller accordingly.
If I remove a specific resource I get in the logs:
```
E1105 15:25:41.372782       1 reflector.go:178] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:224: Failed to list *v1beta1.PingSource: the server could not find the requested resource (get pingsources.sources.knative.dev)
```
which is the same log entry we get in the controller in the eventing namespace. This is acceptable. The user remove the pingsource crd for some reason etc.
In another scenario if the user removes a component eg. Eventing and then re-installs it then he gets in the logs:
```
{"level":"info","ts":"2020-11-05T17:06:38.215Z","logger":"knative.openshift.telemetry","msg":"Starting telemetry for:","component":"Serving"}
{"level":"info","ts":"2020-11-05T17:07:55.466Z","logger":"knative.openshift.telemetry","msg":"Starting telemetry for:","component":"Eventing"}
{"level":"info","ts":"2020-11-05T17:21:22.555Z","logger":"knative.openshift.telemetry","msg":"Stopping telemetry for:","component":"Eventing"}
{"level":"info","ts":"2020-11-05T17:23:50.998Z","logger":"knative.openshift.telemetry","msg":"Starting telemetry for:","component":"Eventing"}
```
No watcher error is emitted in this case.

skonto commented:
/retest

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,92 @@
+	if os.Getenv(SkipTelemetryEnvVar) != "" {
+		return
+	}
+	mu.Lock()
```
This should add a `defer mu.Unlock()` underneath.

Point in case: You currently have a deadlock if you return early in the diff below.

skonto replied:
Good catch, yes.

skonto commented:
Updated the implementation and tested the following scenarios:
a) kill the operator pod and check metrics after the leader comes up. Observed metrics come up consistent with their values before kill.
b) removed eventing or serving crds (uninstalling both) and observed metrics after applying again the corresponding crds and after the components came up.

skonto commented:
/retest

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,180 @@
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func TryStartTelemetry(c controller.Controller, mgr manager.Manager, stop chan struct{}, component string, api client.Client) error {
```
I would've expected this to be part of a `Telemetry` type that you could store on the reconciler itself. You'd then just call `Start` and `Stop` accordingly.

This would solve two issues in the current code:
1. The pointer-replacement shenanigans with the stop channel below.
2. The skip env var becomes unnecessary as you can initialize the reconciler with a noop telemetry object in the test.

skonto replied:
@markusthoemmes I can try that to improve compaction but I dont see them as issues. The reason I followed this is because I didnt want to add a lot of noise in the reconciler anyway.

markusthoemmes replied:
True, they are not issues per-se, but they are unnecessary obstructions imo. Types are cheap. I think this currently adds more noise to the reconcilers than encapsulating the telemetry controllers in a type would.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,180 @@
+		return
+	}
+	log.Info("stopping telemetry for:", "component", component)
+	// Stop the telemetry controller
+	// the lock above makes sure we close once and not panic
+	close(*stop)
+	// Can't use a closed channel
+	*stop = make(chan struct{})
```
This is the shenanigans I refer to above.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,180 @@
+// Serverless Operator is running. No metric label is removed when components are removed so that
+// if any component eg. Eventing is removed and CRDs exist, it will show existing CRs correctly.
+// Telemetry should be used in the background so no error is returned here.
+func InitializeAndTakeMetricsSnapshot(component string, api client.Client) {
```
Is this **really** necessary? I suppose we could have slightly skewed metrics for a reporting period if we don't do this (as the controller does its global resync on startup), but is that slight skew really worth the trouble? :thinking:

skonto replied:
If I create N objects eg 1000 and then restart the operator I get 0. This is a big difference.

markusthoemmes replied:
You get that 0 value until the controller has fully resynced, no?

skonto replied:
What if it does not happen and we get 0 for long?

markusthoemmes replied:
Why would it not happen?

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_ucontroller.go:
```diff
@@ -0,0 +1,113 @@
+
+func (metricsPredicate) Delete(e event.DeleteEvent) bool {
+	if metric := getMetricFor(e.Object); metric != nil {
+		if inSnapshot(e.Meta, getComponentFor(e.Object)) {
```
Why don't we store the snapshot alongside the stopChannel in a `Telemtry` type to not need to do this resolution (`getComponentFor`) at all? This is the cardinality breakup we've talked about.

skonto replied:
Ok see my comment above I will try to do that way.

markusthoemmes commented on knative-operator/pkg/controller/knativeeventing/knativeeventing_controller.go:
```diff
@@ -26,7 +27,14 @@ import (
+var (
+	log           = common.Log.WithName("controller")
+	stopTelemetry = make(chan struct{})
+	// Protects from processing order, if true we should install telemetry
+	// if it is false we need to uninstall it in the next delete stage.
+	// We start by assuming no telemetry is available.
+	shouldInstall = true
```
I don't quite parse the need for this. Again this would be easier if we have a `Telemetry` type encapsulating everything and potentially keeping internal state.

skonto replied:
If you get a delete event first you need to avoid deleting stuff. If there is no chance I can remove it.

markusthoemmes replied:
Sure, but it feels quite odd to carry that state in the reconciler itself, seeing as it belongs to the telemetry controller in question.

skonto commented:
@markusthoemmes I removed the feature flag, now using a noop obj, moved to using a telemetry type. I tested again by killing eventing, serving etc. I am planning to add integration tests for the monitoring stuff in a feature PR.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+	t.InitializeAndTakeMetricsSnapshot(api)
+	// Start our controller in a goroutine so that we do not block.
+	if t.tc != nil { // meant to allow nop objects in tests
```
Shouldn't this be a `if t == nil { return }` at the beginning of the function?

skonto replied:
It could be but it should be a bug if this happens. At least that is the common practice about whether to check pointers in receiver methods.

markusthoemmes replied:
I was actually saying that because we rely on the zero value (i.e. nil) in tests apparently. We use that pattern in other methods on this type already.

skonto replied:
I will change in all functions anyway. I could not parse it above because if I do it for one I should do it for all imho.

markusthoemmes replied:
Yeah I agree, so i didn't parse why we did it different here than in the others.

skonto replied:
Because I read that you should not check the pointer in receiver methods as a practice. Anyway will do it.

markusthoemmes replied:
But you do it already?

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+		return nil, err
+	}
+	t.tc = tc
+	return t, nil
```
```suggestion
	return &Telemetry{
	  name: name,
	  stop: make(chan struct{}),
	  tc: tc
}, nil
```

skonto replied:
Here I need to pass a telemetry ref to the controller, it too late.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+func (t *Telemetry) addToSnaphost(obj metav1.Object) {
```
```suggestion
func (t *Telemetry) addToSnapshot(obj metav1.Object) {
```

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+	}, true)
+}
+
+func (t *Telemetry) deleteFromSnaphost(obj metav1.Object) {
```
```suggestion
func (t *Telemetry) deleteFromSnapshot(obj metav1.Object) {
```

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+	return
```
please don't use anonymous returns. Adding `return ok` makes this a lot clearer imo.

skonto replied:
ok.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_ucontroller.go:
```diff
@@ -0,0 +1,100 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	ServingObjects = []runtime.Object{
+		&servingv1.Service{},
+		&servingv1.Revision{},
+		&servingv1.Route{},
+	}
+
+	EventingObjects = []runtime.Object{
+		&eventingsourcesv1beta1.PingSource{},
+		&eventingsourcesv1beta1.ApiServerSource{},
+		&eventingsourcesv1beta1.SinkBinding{},
+	}
+
+	KnativeKafkaObjects = []runtime.Object{
+		&kafkasourcev1beta1.KafkaSource{},
+	}
+)
```
These shouldn't live here but in the respectivecontrollers.

skonto replied:
ok.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+func (t *Telemetry) TryStopTelemetry() {
+	log.Info("stopping telemetry for:", "component", t.name)
+	// Stop the telemetry controller
+	// the lock above makes sure we close once and not panic
```
Which lock?

skonto replied:
Doc needs update.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_ucontroller.go:
```diff
@@ -0,0 +1,100 @@
+
+var (
+	ServingObjects = []runtime.Object{
+		&servingv1.Service{},
+		&servingv1.Revision{},
+		&servingv1.Route{},
```
Why no `Configuration`?

skonto replied:
I thought the same, in SRVKS mentions only the above, so from a telemetry perspective may not be useful. I can add it if it useful for telemetry.

skonto replied:
I will add it, should be of no harm.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry_ucontroller.go:
```diff
@@ -0,0 +1,100 @@
+// newTelemetryController creates an unmanaged controller for watching Telemetry resources
+func newTelemetryController(name string, objects []runtime.Object, mgr manager.Manager, telemetry *Telemetry) (*controller.Controller, error) {
+	// Create a new controller
+	c, err := controller.NewUnmanaged(fmt.Sprintf("telemetry-resources-%s-controller-%s", name, uuid.New().String()), mgr, controller.Options{
```
What do we gain by adding the uuid in here?

skonto replied:
Different names in logs.

markusthoemmes replied:
Do we do that for our other controllers? I'd actually not want to do that as the strings should actually be stable to allow for easy filtering across pods.

markusthoemmes commented on knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:
```diff
@@ -38,6 +40,10 @@ var (
+	// Protects from processing order, if true we should install telemetry
+	// if it is false we need to uninstall in the next delete stage.
+	// We start by assuming no telemetry is available.
+	shouldInstallTelemetry = true
```
This should move inside the telemetry implementation.

skonto replied:
It could be but it is also applied to the current reconciler.

markusthoemmes commented on knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:
```diff
@@ -96,6 +103,12 @@ func add(mgr manager.Manager, r *ReconcileKnativeKafka) error {
 		}
 	}
 
+	t, err := telemetry.NewTelemetry("knativeKafka", mgr, telemetry.KnativeKafkaObjects)
+	if err != nil {
+		return err
+	}
+	r.telemetry = t
```
This should be part of `newReconciler`

skonto replied:
ok.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,186 @@
+	case "eventing":
+		sourcesG = serverlessTelemetryG.WithLabelValues("source")
+		pingSourceList := &eventingsourcesv1beta1.PingSourceList{}
+		if err := api.List(context.TODO(), pingSourceList); err == nil {
```
I wonder if we should just always use a List operation to gather metrics then. AFAIK, they should be coming from cache anyway (with the `Watch` being setup), so what if we don't do an `Add/Dec` operation on individual changes but instead do a `Set` operation on each individual change by listing everything. If everything comes from cache, there's 0 cost to that and it'd make the "snapshot" obsolete.

skonto replied:
The problem is that for stuff that are not in cache yet you will do that call multiple times for each type of resource you get an event for. The reconciler uses c-r client to list stuff, and it touches the cache as you mention. 
The other problem is that if you have no event coming in to trigger the controller because someone restarted the operator pod you will get a zero value for all metrics. T

markusthoemmes replied:
But shouldn't we get a global resync on controller startup?

skonto replied:
The complete solution would be to do a global resync there as well. In the code I have this comment:

```
	// Initialize metrics before we start the corresponding controller.
	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
``
At this point we take the snapshot and then start the controller. I dont expect to lose a considerable number of events in between.
For telemetry I am going to use something like `avg by (type) (avg_over_time(serverless_telemetry[30m]))`, so I dont expect much of a difference in general. I can also do that for the sake of correctness. WDYTH?

markusthoemmes replied:
My point is: Why doesn't it already do a global resync anyway? AFAIK starting a watcher does that by design.

skonto replied:
Ok will do.

skonto commented:
@markusthoemmes I moved to a better design I update state per metric, also in the controller. No coordination is required as metrics match 1-1 to the events. Metrics are as follows:
![image](https://user-images.githubusercontent.com/7945591/98596772-9acbd380-22e0-11eb-8605-33b1aae36dbe.png)

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,121 @@
+		log.Info("stopping telemetry for:", "component", t.name)
+		// Stop the telemetry controller
+		close(t.stop)
+		// Can'telemetry use a closed channel
```
```suggestion
		// Can't reuse a closed channel
```

skonto commented:
/retest

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,121 @@
+		case err := <-errCh:
+			t.shouldInstallTelemetry = true
+			return err
+		case <-time.After(1 * time.Second): // assume no error, as controller has blocked
```
This will randomly block the reconcile loop for 1 second :thinking:. Should we "just" resort on error logging instead?

skonto replied:
I am trying to find some way to extract the error. I can remove it.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,121 @@
+	if t == nil {
+		return
+	}
+	switch t.name {
+	case "eventing":
+		pingSourceG = serverlessTelemetryG.WithLabelValues("source_ping")
+		apiServerSourceG = serverlessTelemetryG.WithLabelValues("source_apiserver")
+		sinkBindingSourceG = serverlessTelemetryG.WithLabelValues("source_sinkbinding")
+		updateMetricFor(&eventingsourcesv1beta1.ApiServerSource{}, api)
+		updateMetricFor(&eventingsourcesv1beta1.PingSource{}, api)
+		updateMetricFor(&eventingsourcesv1beta1.SinkBinding{}, api)
+	case "knativeKafka":
+		kafkaSourceG = serverlessTelemetryG.WithLabelValues("source_kafka")
+		updateMetricFor(&kafkasourcev1beta1.KafkaSource{}, api)
+	case "serving":
+		serviceG = serverlessTelemetryG.WithLabelValues("service")
+		routeG = serverlessTelemetryG.WithLabelValues("route")
+		revisionG = serverlessTelemetryG.WithLabelValues("revision")
+		configurationG = serverlessTelemetryG.WithLabelValues("configuration")
+		updateMetricFor(&servingv1.Service{}, api)
+		updateMetricFor(&servingv1.Route{}, api)
+		updateMetricFor(&servingv1.Revision{}, api)
+		updateMetricFor(&servingv1.Configuration{}, api)
+	}
```
Are these lists redundant with the constructor?

I.e. could we store the objects passed and then do:

```golang
for _, obj := t.objects {
  updateMetricFor(obj, api)
}
```

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,121 @@
+	}
+	switch t.name {
+	case "eventing":
+		pingSourceG = serverlessTelemetryG.WithLabelValues("source_ping")
```
What are these label values for?

skonto replied:
It is a gauge vec where you have one vec and set a label to create a number of different time series based on the labels. It is standard prometheus approach instead of creating N different vecs.

markusthoemmes replied:
Ah I see, and we can't initialize them in the metrics code becauuuuse... that would make them all show 0 and we don't want that?

skonto replied:
Yes at least at the beginning we dont want if some component is never used or the operator is restarted, but it is of no harm to initialize them (I had it that way before). 
However, when we stop the controller we could remove the labels by calling `serverlessTelemetryG.DeleteLabelValues("...")`, I am still thinking about it, as discussed [here](https://godoc.org/github.com/prometheus/client_golang/prometheus#MetricVec.GetMetricWithLabelValues).:
```
Keeping the Metric for later use is possible (and should be considered if performance is critical), but keep in mind that Reset, DeleteLabelValues and Delete can be used to delete the Metric from the MetricVec. In that case, the Metric will still exist, but it will not be exported anymore, even if a Metric with the same label values is created later.
```
Right now between deletes/installs of Serving, Eventing while the serverless operator is running we keep the metrics.

markusthoemmes replied:
Honestly, I would not want to do that. What's the harm in reporting 0 sources in a cluster where there are... well... 0 sources? Seems like unnecessary complexity to me.

markusthoemmes reviewed (commented):
The cardinality you show in the screenshot is a showstopper for Telemetry I believe, as this will be different for each cluster :thinking:. Have we checked with them if this is okay?

skonto commented:
@markusthoemmes for telemetry we will have aggregation over these metrics.

skonto commented:
/retest

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,104 @@
+		t.initializeAndUpdateMetrics(api)
+		// Start our controller in a goroutine so that we do not block.
+		go func() {
+			// Block until our controller manager is elected leader. We presume our
+			// entire process will terminate if we lose leadership, so we don'telemetry need
+			// to handle that.
+			<-mgr.Elected()
```
I think we can safely drop this, can't we? We should be guaranteed to be leaders by the time we call Start in the first place.

skonto replied:
Yes we can.

markusthoemmes commented on knative-operator/pkg/common/telemetry/telemetry.go:
```diff
@@ -0,0 +1,104 @@
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func (t *Telemetry) TryStartTelemetry(api client.Client, mgr manager.Manager) error {
```
```suggestion
func (t *Telemetry) TryStart(api client.Client, mgr manager.Manager) error {
```

and same for Stop, to have a bit less stutter in the API.

skonto replied:
A bit nit but ok.

markusthoemmes replied:
Wouldn't have commented if there wasn't more in the review :P

skonto commented:
> Honestly, I would not want to do that. What's the harm in reporting 0 sources in a cluster where there are... well... 0 sources? Seems like unnecessary complexity to me.

Ok.

>  Do we do that for our other controllers? I'd actually not want to do that as the strings should actually be stable to allow for easy filtering across pods.
I can remove it.

skonto commented:
@markusthoemmes I will follow up this with an integration test, let me know if it is now ok.

markusthoemmes reviewed (approved):
/lgtm
/approve
